sdk/containerregistry/container-registry/src/blob/containerRegistryBlobClient.ts

```diff
@@ -0,0 +1,410 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  InternalPipelineOptions,
+  bearerTokenAuthenticationPolicy,
+  RestError,
+} from "@azure/core-rest-pipeline";
+import { TokenCredential } from "@azure/core-auth";
+import { GeneratedClient } from "../generated";
+import { ChallengeHandler } from "../containerRegistryChallengeHandler";
+import { ContainerRegistryRefreshTokenCredential } from "../containerRegistryTokenCredential";
+import { logger } from "../logger";
+import { calculateDigest } from "../utils/digest";
+import {
+  DeleteBlobOptions,
+  DeleteManifestOptions,
+  DownloadBlobOptions,
+  DownloadBlobResult,
+  DownloadManifestOptions,
+  DownloadManifestResult,
+  OciManifest,
+  UploadBlobResult,
+  UploadManifestOptions,
+  UploadManifestResult,
+} from "./models";
+import * as Mappers from "../generated/models/mappers";
+import { CommonClientOptions, createSerializer } from "@azure/core-client";
+import { readStreamToEnd } from "../utils/helpers";
+import { Readable } from "stream";
+import { createSpan } from "../tracing";
+import { SpanStatusCode } from "@azure/core-tracing";
+
+const LATEST_API_VERSION = "2021-07-01";
+
+enum KnownManifestMediaType {
+  OciManifestMediaType = "application/vnd.oci.image.manifest.v1+json",
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+function withRequiredProperty<T, U extends keyof T>(property: U, obj: T): T & Required<Pick<T, U>> {
+  if (!Object.prototype.hasOwnProperty.call(obj, property)) {
+    throw new RestError(`Expected property ${property} to be defined.`);
+  }
+
+  return obj as T & Required<Pick<T, U>>;
+}
```

Comment:
Small preference for 

```suggestion
function assertRequiredProperty<T, U extends keyof T>(property: U, obj: T): asserts obj is T & Required<Pick<T, U>> {
  if (!Object.prototype.hasOwnProperty.call(obj, property)) {
    throw new RestError(`Expected property ${property} to be defined.`);
  }
}
```

so you can use it like

```ts
const result = await this.client.containerRegistry.createManifest(
    this.repositoryName,
    tagOrDigest,
    manifestBody,
    { contentType: KnownManifestMediaType.OciManifestMediaType, ...updatedOptions }
  )
);
assertRequiredProperty("dockerContentDigest", result);

return { digest: result.dockerContentDigest };
```

Mostly because nesting the 'assert' inside of a method call seems like an easy way to forget to do it later, and then we're asserting on a Promise object instead of the real result.